search-service/src/main/kotlin/com/egm/stellio/search/csr/model/ContextSourceRegistration.kt

```diff
@@ -0,0 +1,132 @@
+                relationshipNames = relationshipNames?.map { expandJsonLdTerm(it, contexts) },
+            )
+
+        fun compact(contexts: List<String>): RegistrationInfo =
+            RegistrationInfo(
+                entities = entities?.map { it.compact(contexts) },
+                propertyNames = propertyNames?.map { compactTerm(it, contexts) },
+                relationshipNames = relationshipNames?.map { compactTerm(it, contexts) }
+            )
+    }
+
+    data class EntityInfo(
+        val id: URI?,
+        val idPattern: String?,
+        val type: List<String> // todo can be String
+    ) {
+        fun expand(contexts: List<String>): EntityInfo =
+            this.copy(
+                type = type.map { expandJsonLdTerm(it, contexts) },
+            )
+
+        fun compact(contexts: List<String>): EntityInfo =
+            this.copy(
+                type = type.map { compactTerm(it, contexts) },
+            )
+    }
+
+    fun expand(contexts: List<String>): ContextSourceRegistration =
+        this.copy(
+            information = information.map { it.expand(contexts) }
+        )
+
+    fun compact(contexts: List<String>): ContextSourceRegistration =
+        this.copy(
+            information = information.map { it.compact(contexts) }
+        )
+
+    fun serialize(
+        contexts: List<String>,
+        mediaType: MediaType = JSON_LD_MEDIA_TYPE,
+        includeSysAttrs: Boolean = false
+    ): String {
+        val representation = NgsiLdDataRepresentation(
+            entityRepresentation = EntityRepresentation.forMediaType(mediaType),
+            includeSysAttrs = includeSysAttrs,
+            attributeRepresentation = AttributeRepresentation.NORMALIZED
+        )
+        return mapper.writeValueAsString(
+            mapper.convertValue<Map<String, Any>>(
+                (this.compact(contexts))
+            ).toFinalRepresentation(representation) // todo dangerous?
+        )
+    }
+
+    companion object {
+        fun deserialize(
+            input: Map<String, Any>,
+            contexts: List<String>
+        ): Either<APIException, ContextSourceRegistration> =
+            runCatching {
+                deserializeAs<ContextSourceRegistration>(serializeObject(input.plus(JSONLD_CONTEXT to contexts)))
+                    .expand(contexts)
+            }.fold(
+                { it.right() },
+                { it.toAPIException("Failed to parse context_source_registration").left() }
+            )
+
+        fun contextSourceRegistrationNotFoundMessage(id: URI) = "Could not find a CSourceRegistration with id $id"
+        fun contextSourceRegistrationAlreadyExistsMessage(id: URI) = "A CSourceRegistration with id $id already exists"
+        fun contextSourceRegistrationUnauthorizedMessage(id: URI) =
+            "User is not authorized to access CSourceRegistration $id"
+        const val TYPE = "ContextSourceRegistration"
```

Comment:
`CSR_TYPE` to make it more explicit when used elsewhere

Reply:
seems duplicate of `NGSILD_CSR_TERM` defined in `JsonLdUtils` (used when checking type at creation time)